### DISCUSSION TITLE
Is there a way to change an injected (test)value after initalization?

### DISCUSSION THREAD

**Discussion by agrabz** [pointfreeco/swift-dependencies#154] [Q&A] [answered]
I know that it sounds a bad and non-clean idea, but what I want to test is the error handling of a list refreshing logic with the below steps:

1. Load values into a list **successfully**
2. Try refresh (the action called by pull to refresh) which will **fail**

As both steps are using the same logic (through the same dependency, let's call it `DataFetcher`) I couldn't figure out a way to test this with Swift Dependencies. See example code with comments below:

```
func test_Event_Refresh_Failure() async {
        await withMainSerialExecutor {
            //GIVEN
            let sut = withDependencies {
                $0.dataFetcher = DataFetcher()
            } operation: {
                ListViewModel()
            }
            
            await sut.handleAsyncEvent(.viewIsInitialized) //calls DataFetcher.fetchList(), should succeed
            
            //WHEN
           //would be nice if I could change the value of DataFetcher to MockDataFetcherWithError() before this line.
            let task = Task { await sut.handleAsyncEvent(.didPullToRefresh) } //calls DataFetcher.fetchList(), should fail to test error handling logic
            await Task.yield()
            //THEN
            XCTAssert(sut.isListRefreshing == true)
            await task.value
            XCTAssert(sut.isListRefreshing == false)
            XCTAssert(sut.isListInError == true)
        }
    }
```

I guess I should rather change my code to handle this differently, so I'm also open to receive any kind of feedback for changing this solution to a more testable implementation, but if there's anyway to change the value after init, that'd be great!

**Comment by mbrandonw** [comment, marked as answer]
Hi @agrabz, you can simply use `withDependencies` again to alter the dependency environment for a well-defined scope:

```swift
//WHEN
//would be nice if I could change the value of DataFetcher to MockDataFetcherWithError() before this line.
await withDependencies {
  $0.dataFetch = MockDataFetcherWithError()
} operation: {
  let task = Task { await sut.handleAsyncEvent(.didPullToRefresh) } //calls DataFetcher.fetchList(), should fail to test error handling logic
  await Task.yield()
  //THEN
  XCTAssert(sut.isListRefreshing == true)
  await task.value
  XCTAssert(sut.isListRefreshing == false)
  XCTAssert(sut.isListInError == true)
}
```

**Reply by agrabz** [reply]
Thanks a lot @mbrandonw ! That was so obvious. I misinterpreted the docs and thought that withDependencies is used only for object initialisation.
